### PR TITLE
New (imo improved) artist parsing

### PR DESCRIPTION
**Pull request by TcMaX**
Main functionality is removing regex and relying on native functions. Beyond that, is a fair bit more rigorous (for instance, if there was an artist with a dash in their name, say blume - popo, "blume - popo -- title" would be interpreted as blume, not blume - popo. There are a couple more improvements as well, though I mightve missed some functionality due to my inexperience with regex). Does take up more space though, though it might be faster simply because regex. Didn't benchmark though.

The changes to the return mechanism is necessary for the functioning of the song parsing I commited in #14 (though it won't work without some changes, so this should be merged before #14, whether in its current form or with the regex changes reverted, but return mechanism preserved)

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -113,21 +115,68 @@ def _remove_submission_exceeding_threshold(self, submission, service):
...
-            # If there is a featuring artists, it should use the feature_artist artist result
-            return feature_artist.group('artist')
-        if artist is not None and artist.group('artist') is not None:
-            return artist.group('artist')
 
+        #get main artist
+        dashChar = ["-","—"]
+        doubleDash = False
+        artist = None
+        for x in dashChar:
```
[the remark this answers is not in the document]
Comment (martijnboers) [reply main.py]:
```suggestion
        for x in ["-" , "—"]:
```
codestyle: when using a variable only once I prefer in-lining it

File: main.py [hunk 2]
```diff
@@ -113,21 +115,68 @@ def _remove_submission_exceeding_threshold(self, submission, service):
...
+            if((x+x) in post_title):
+                artist = post_title.split(x+x)[0].strip()
+                doubleDash = True
+                break
+        
+        if(doubleDash is False):
```
Comment (martijnboers) [inline main.py]:
```suggestion
        if doubleDash is False:
```

File: main.py [hunk 3]
```diff
@@ -113,21 +115,68 @@ def _remove_submission_exceeding_threshold(self, submission, service):
...
-        feature_artist = re.search(
-            r'(?P<artist>.+?)\s*(&|feat.?|ft\.?|feature|featuring)\s*(?P<featureArtist>.+?)\s*(-|—)\s*',
-            post_title, re.IGNORECASE
-        )
-
-        if feature_artist is not None:
-            # If there is a featuring artists, it should use the feature_artist artist result
-            return feature_artist.group('artist')
-        if artist is not None and artist.group('artist') is not None:
-            return artist.group('artist')
```
Comment (martijnboers) [inline main.py]:
```suggestion
```

File: main.py [hunk 4]
```diff
@@ -113,21 +115,68 @@ def _remove_submission_exceeding_threshold(self, submission, service):
...
+        artist = None
+        for x in dashChar:
+            if((x+x) in post_title):
+                artist = post_title.split(x+x)[0].strip()
+                doubleDash = True
+                break
+        
+        if(doubleDash is False):
+            for x in dashChar:
+                if(x in post_title):
```
Comment (martijnboers) [inline main.py]:
```suggestion
                if x in post_title:
```

**Comment by martijnboers** [comment]
@TcMaX I added you as a collaborator so you can create branches on this repo vs your own fork

**Comment by TcMaX** [comment]
Fixed most of the things requested for formatting and naming now. Going to get around to the rest of the stuff later today probably.

File: main.py [hunk 5]
```diff
@@ -113,21 +115,68 @@ def _remove_submission_exceeding_threshold(self, submission, service):
     def _reply_with_sticky_post(self, submission, reply_text):
         comment = submission.reply(reply_text)
         comment.mod.distinguish("yes", sticky=True)
-
+    
+    def _get_prioritized_artist(artistList):
```
[the remark this answers is not in the document]
Comment (TcMaX) [reply main.py]: Will do

**Comment by TcMaX** [comment]
If I erronously resolved a comment just recomment if the issue is still there

File: main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TcMaX) [reply main.py]: This particular one was not used just once, but the two others that were have been inlined

File: main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TcMaX) [reply main.py]: Done

**Comment by TcMaX** [comment]
I think I updated the test file properly, though I cant really get that test file to work personally, but I couldnt before either, I think im just incompetent. Anyway idk how to move this branch to main, but I'll keep in mind to branch this one rather than fork next time Im branching

File: title_parser.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+        for feature in ['feat.', 'ft.', 'feature', 'featuring', '&']:
+            if feature in lower_title:
+                feat_index = lower_title.index(x)
+
+                # remove featuring from artist if exists
+                if len(artist) > feat_index:
+                    artist = artist[:feat_index].strip()
+
+                feat_index += len(x)
```
Comment (martijnboers) [inline title_parser.py]: `x` is undefined

File: title_parser.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+                # isolate featuring artist
+                feature_artist = post_title[featIndex:].strip()
+                break
+
+        # further process if found
+        if feature_artist is not None:
+            for end_char in [' -', ')', '[', ' —']:
+                if end_char in feature_artist:
+                    feature_artist = feature_artist.split(x)[0]
```
Comment (martijnboers) [inline title_parser.py]: `x` is undefined

**Comment by TcMaX** [comment]
Ended up doing this idk. Anyway I'm gonna submit in your format now, though im going to add those two extra tests, to ensure ー is never added as a dash character, and to ensure checking for double dash priority remains. Commiting that with some fixes to parser in a moment

`from unittest import TestCase

from title_parser import TitleParser


class TestBlotterTrax(TestCase):
    @staticmethod
    def test_extract_artist_post_title():
        print("yes")
        submissions = [
            ('B12 - Infinite Lites (Original Mix) ', 'B12', None),
            ('Floating Points - Last Bloom ', 'Floating Points', None),
            ('Empty Spaces -- Một Cuộc Sống Khác [Vietnamese/Alternative] (2019) ', 'Empty Spaces', None),
            ('007Bonez & Adro - Motion [Hip-Hop / Rap] (2019)', '007Bonez', 'Adro'),
            ('007Bonez featuring Adro - Motion [Hip-Hop / Rap] (2019)', '007Bonez', 'Adro'),
            ('007Bonez feat Adro - Motion [Hip-Hop / Rap] (2019)', '007Bonez feat Adro', None),
            ('007Bonez feat. Adro - Motion [Hip-Hop / Rap] (2019)', '007Bonez', 'Adro'),
            ('Teen Suicide — Haunt Me x3 [indie rock] (2014)', 'Teen Suicide', None),
            ("upsammy - Another Place - Nous'klaer 011", 'upsammy', None),
            ('ガールズロックバンド革命 - CHANGE', 'ガールズロックバンド革命', None),
            ('Blume - popo -- 溺レル', 'Blume - popo', None)
        ]
        for submission_title, artist, featuring_artist in submissions:
            title = TitleParser._get_artist_name_from_submission_title(submission_title)

            print(title[0] == artist)
            print(title[1] == featuring_artist)

TestBlotterTrax.test_extract_artist_post_title()`

**Pull request merged**